### DISCUSSION TITLE
Unable to load type System.Drawing.Bitmap required for deserialization.

### DISCUSSION THREAD

**Discussion by dayAndnight2018** [AvaloniaUI/Avalonia#14647] [Q&A]
Unable to load type System.Drawing.Bitmap required for deserialization. It hits exception when I invoke the method to fetch image from the clipboard on window 10 desktop.

this.Clipboard.GetDataAsync("System.Drawing.Bitmap")

**Comment by dayAndnight2018** [comment]
The image is put in one winform app by method 

Clipboard.SetData("System.Drawing.Bitmap", new Bitmap());

**Comment by maxkatz6** [comment]
Include System.Drawing.Common nuget package, if you need to deserialize data this way.
But I would advise to not rely on .NET BinaryFormatter (which is used when you use .NET type formats on Windows clipboard), as it's going to be removed from .NET 9.

**Reply by dayAndnight2018** [reply]
But it does not work when I and system.drawing.common package

**Reply by timunie** [reply]
#6501 https://github.com/AvaloniaUI/Avalonia/pull/7329

**Reply by dayAndnight2018** [reply]
By the way, Where is the implements of the method ReadImageHeader?

**Reply by dayAndnight2018** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/44399091/adb3bb48-2c4f-47e7-93bd-02376d53368b)

**Reply by dayAndnight2018** [reply]
The pixelsData is null from unknown_format_2

**Reply by timunie** [reply]
Probably. Depends on the data in the clipboard iirc. I'm not that deep into that area unfortunately. There are tool to spy the clipboard on windows which you can use to check out what is actually there.

**Reply by timunie** [reply]
https://delphidabbler.com/software/cfs for example

**Reply by dayAndnight2018** [reply]
solve it!

`if (format.Contains("Unknown_Format_8"))
                    {
                        var imagePixelsData = await this.Clipboard.GetDataAsync("Unknown_Format_8") as byte[];
                        var bitmapInfoData = await this.Clipboard.GetDataAsync("Unknown_Format_8") as byte[];
                        var header = ReadBitmapInfo(bitmapInfoData);

                        var pixels = new SKColor[header.biWidth * header.biHeight];
                        var pixelIndex = header.biWidth * header.biHeight - 1;

                        var pos = 52; // length of bitmap header 
                        for (var y = header.biHeight - 1; y >= 0 ; y--)
                        {
                            for (var x = header.biWidth - 1; x >= 0 ; x--)
                            {
                                var b = imagePixelsData[pos++];
                                var g = imagePixelsData[pos++];
                                var r = imagePixelsData[pos++];
                                if (header.biBitCount == 32)
                                {
                                    var a = imagePixelsData[pos++];
                                    pixels[y * header.biWidth + (header.biWidth - 1 - x)] = new SKColor(r, g, b, a);
                                    continue;
                                }
                                pixels[pixelIndex++] = new SKColor(r, g, b);
                            }
                        }

                        var bitmap = new SKBitmap(new SKImageInfo(header.biWidth, header.biHeight));
                        bitmap.Pixels = pixels;

                        var path = "xxx";
                        var fileName = "yyy";
                        
                        using var writeStream = File.Open(Path.Combine(path, fileName), FileMode.OpenOrCreate, FileAccess.Write);
                        bitmap.Encode(SKEncodedImageFormat.Png, 100).SaveTo(writeStream);
                    }`

**Reply by dayAndnight2018** [reply]
thanks a lot!

**Comment by dayAndnight2018** [comment]
> https://delphidabbler.com/software/cfs for example



> https://delphidabbler.com/software/cfs for example

**Comment by dayAndnight2018** [comment]
public class BitmapHeaderInfo
    {
        public uint biSize;
        public int biWidth;
        public int biHeight;
        public ushort biPlanes;
        public ushort biBitCount;
        public uint biCompression;
        public uint biSizeImage;
        public int biXPelsPerMeter;
        public int biYPelsPerMeter;
        public uint biClrUsed;
        public uint biClrImportant;
    }

**Comment by dayAndnight2018** [comment]
private BitmapHeaderInfo ReadBitmapInfo(byte[] data)
        {
            using var memStream = new MemoryStream(data);
            using var reader = new BinaryReader(memStream);
            return new BitmapHeaderInfo
            {
                biSize = reader.ReadUInt32(),
                biWidth = reader.ReadInt32(),
                biHeight = reader.ReadInt32(),
                biPlanes = reader.ReadUInt16(),
                biBitCount = reader.ReadUInt16(),
                biCompression = reader.ReadUInt32(),
                biSizeImage = reader.ReadUInt32(),
                biXPelsPerMeter = reader.ReadInt32(),
                biYPelsPerMeter = reader.ReadInt32(),
                biClrUsed = reader.ReadUInt32(),
                biClrImportant = reader.ReadUInt32()
            };
        }
